File: nodekit/_internal/types/actions.py
from abc import ABC
from typing import Literal, Union, Annotated, Dict, Any

import pydantic

from nodekit._internal.types.values import PressableKey, SpatialPoint, Value
from typing import TypedDict

# %%
class BaseAction(pydantic.BaseModel, ABC):
    action_type: str
    action_value: Any



# %%
class KeyAction(BaseAction):
    action_type: Literal["KeyAction"] = "KeyAction"
    key: PressableKey = pydantic.Field(description="The key that was pressed.")


# %%
class SliderAction(BaseAction):
    action_type: Literal["SliderAction"] = "SliderAction"
    bin_index: int = pydantic.Field(
        description="The index of the bin that was selected.", ge=0
    )


# %%
class TextEntryAction(BaseAction):
    action_type: Literal["TextEntryAction"] = "TextEntryAction"
    text: str


# %%
class WaitAction(BaseAction):
    action_type: Literal["WaitAction"] = "WaitAction"


# %%
class SelectAction(BaseAction):
    action_type: Literal["SelectAction"] = "SelectAction"
    selection: str


# %%
class MultiSelectAction(BaseAction):
    action_type: Literal["MultiSelectAction"] = "MultiSelectAction"
    selections: list[str]


# %%
class ProductAction(BaseAction):
    action_type: Literal["ProductAction"] = "ProductAction"
    child_actions: Dict[str, "Action"]


# %%
class SumAction(BaseAction):
    action_type: Literal["SumAction"] = "SumAction"
    child_id: str
    child_action: "Action"


# %%
type Action = Annotated[
    Union[
        KeyAction,
        SliderAction,
        TextEntryAction,
        WaitAction,
        SelectAction,
        MultiSelectAction,
        ProductAction,
        SumAction,
    ],
    pydantic.Field(discriminator="action_type"),
]

File: nodekit-browser/src/types/actions.ts
import type {Dict, PressableKey} from "./values.ts";
import type {SliderBinIndex} from "../board-view/sensor-bindings/slider";

export interface BaseAction<T extends string> extends Dict {
    action_type: T
    action_value: any
}

export interface KeyAction extends BaseAction<"KeyAction"> {
    action_value: PressableKey;
}

export interface SliderAction extends BaseAction<"SliderAction"> {
    action_value: SliderBinIndex
}
export interface TextEntryAction extends BaseAction<"TextEntryAction"> {
    action_value: string
}

export interface WaitAction extends BaseAction<"WaitAction">{
    action_value: null;
}

export interface SelectAction extends BaseAction<"SelectAction">{
    action_value: string
}

export interface MultiSelectAction extends BaseAction<"MultiSelectAction">{
    action_value: string[]
}

export interface ProductAction extends BaseAction<'ProductAction'> {
    action_value: Record<string, Action>
}

export interface SumAction extends BaseAction<'SumAction'>{
    action_value: [string, Action]
}

// Union
export type Action =
    | KeyAction
    | SliderAction
    | TextEntryAction
    | WaitAction
    | SelectAction
    | MultiSelectAction
    | ProductAction
    | SumAction

